src/aquarium-optimized/dawn/BufferManagerDawn.cpp

```diff
@@ -88,36 +90,39 @@ BufferManagerDawn::~BufferManagerDawn()
 }
 
 // Allocate new buffer from buffer pool.
-RingBufferDawn *BufferManagerDawn::allocate(size_t size, bool sync)
+RingBufferDawn *BufferManagerDawn::allocate(size_t size, size_t *offset)
 {
-    // TODO(yizhou): Refactor ring buffer allocation. Get the last one and check if the ring buffer
-    // is full. If the buffer can hold extra size space, use the last one directly.
     // If update data by sync method, create new buffer to upload every frame.
     // If updaye data by async method, get new buffer from pool if available. If no available
-    // buffer, create a new buffer.
-    mSync = sync;
+    // buffer and size is enough in the buffer pool, create a new buffer. If size reach the
+    // limit of the buffer pool, force wait for the buffer on mapping.
+    // Get the last one and check if the ring buffer is full. If the buffer can hold extra
+    // size space, use the last one directly.
+    // TODO(yizhou): Return nullptr if size reach the limit or no available buffer, this means
+    // small bubbles in some of the ring buffers and we haven't deal with the problem now.
 
     RingBufferDawn *ringBuffer = nullptr;
-    if (sync)
+    size_t cur_offset          = 0;
+    if (mSync)
     {
+        // Upper limit
         if (mUsedSize + size > mBufferPoolSize)
         {
+            std::cout << "Upper limit" << std::endl;
             return nullptr;
         }
 
         ringBuffer = new RingBufferDawn(this, size);
         mEnqueuedBufferList.emplace_back(ringBuffer);
     }
-    else
+    else  // Buffer mapping async
     {
         while (!mMappedBufferList.empty())
         {
             ringBuffer = static_cast<RingBufferDawn *>(mMappedBufferList.front());
-            mMappedBufferList.pop();
-            if (ringBuffer->getSize() < size)
+            if (ringBuffer->getAvailableSize() < size)
             {
-                mUsedSize += ringBuffer->getSize();
-                ringBuffer->destory();
+                mMappedBufferList.pop();

```

Comment:
I haven't understood the whole picture here. Why to pop the buffer when it's almost full? What will happen to the data in this buffer?

Reply:
MappedBufferLists contains ring buffers have available space. If the first buffer is full, pop it(not destroy) and add it to enquedBufferList, and then use the next empty buffer. Commit all of the staging buffer in enquedBufferList to update all of the data of a frame, and then remap the buffers, and they will be recycled to MappedBufferList again.

Reply:
I see, thanks